pymongo/network_layer.py

```diff
@@ -35,15 +39,65 @@
 BLOCKING_IO_ERRORS = (BlockingIOError, *ssl_support.BLOCKING_IO_ERRORS)
 
 
-async def async_sendall(socket: Union[socket.socket, _sslConn], buf: bytes) -> None:
-    timeout = socket.gettimeout()
-    socket.settimeout(0.0)
+async def async_sendall(sock: Union[socket.socket, _sslConn], buf: bytes) -> None:
+    timeout = sock.gettimeout()
+    sock.settimeout(0.0)
     loop = asyncio.get_event_loop()
     try:
-        await asyncio.wait_for(loop.sock_sendall(socket, buf), timeout=timeout)  # type: ignore[arg-type]
+        if isinstance(sock, (SSLSocket, _sslConn)):
+            if sys.platform == "win32":
+                await asyncio.wait_for(_async_sendall_ssl_windows(sock, buf), timeout=timeout)
+            else:
+                await asyncio.wait_for(_async_sendall_ssl(sock, buf, loop), timeout=timeout)
+        else:
+            await asyncio.wait_for(loop.sock_sendall(sock, buf), timeout=timeout)
     finally:
-        socket.settimeout(timeout)
+        sock.settimeout(timeout)
 
 
-def sendall(socket: Union[socket.socket, _sslConn], buf: bytes) -> None:
-    socket.sendall(buf)
+async def _async_sendall_ssl(
+    sock: Union[socket.socket, _sslConn], buf: bytes, loop: AbstractEventLoop
+) -> None:
+    fd = sock.fileno()
+    sent = 0
+
+    def _is_ready(fut: Future) -> None:
+        if fut.done():
```

Comment:
Should we remove reader/writer before checking done()?

Reply:
Good catch.